File: day3/day3.cpp
#include <fstream>
#include <numeric>
#include <utility>
#include <vector>
#include <fmt/format.h>

using namespace std;

struct Map
{
	vector<string> rows;
	unsigned width;
	unsigned height;

	unsigned count_trees(int dx, int dy) const
	{
		unsigned count = 0;
		unsigned x = 0;
		unsigned y = 0;
		for (;;)
		{
			x = (x + dx) % width;
			y = y + dy;
			if (y >= height)
			{
				break;
			}

			if (rows[y][x] == '#')
			{
				count++;
			}
		}
		return count;
	}
};

istream& operator>>(istream& input, Map& map)
{
	map.width = map.height = 0;
	map.rows.clear();
	string line;
	while (getline(input, line))
	{
		if (line.size()> 0)
		{
			map.width = line.size();
			map.rows.emplace_back(move(line));
		}
	}
	map.height = map.rows.size();
	return input;
}

int main(int argc, char *argv[])
{
	if (argc < 2)
	{
		fmt::print(stderr, "Usage: {} <filename>\n", argv[0]);
		return 1;
	}

	ifstream input(argv[1]);
	if (!input)
	{
		fmt::print(stderr, "Cannot open {}\n", argv[1]);
		return 1;
	}

	Map map;
	input >> map;
	input.close();

	fmt::print("Part1: {}\n", map.count_trees(3, 1));

	const pair<int, int> SLOPES[] = {
		make_pair(1, 1),
		make_pair(3, 1),
		make_pair(5, 1),
		make_pair(7, 1),
		make_pair(1, 2),
	};
	fmt::print(
		"Part2: {}\n",
		accumulate(
			SLOPES,
			SLOPES + sizeof(SLOPES)/sizeof(SLOPES[0]),
			1ULL,
			[&map](unsigned long long a, const pair<int,int>& p) {
				return a * map.count_trees(p.first, p.second);
			}));
	return 0;
}

File: day3/day3.py
#!/usr/bin/env python3

from functools import reduce
import operator
import sys

def make_map(txt):
    return txt.strip().split("\n")

def count_trees(mymap, slope):
    height = len(mymap)
    width = len(mymap[0])
    x, y = 0, 0
    trees = 0
    while True:
        x = (x + slope[0]) % width
        y += slope[1]
        if y >= height:
            break
        if mymap[y][x] == "#":
            trees += 1

    return trees

def mul_slopes(mymap, slopes):
    return reduce(operator.mul, [count_trees(mymap, slope) for slope in slopes])


if __name__ == "__main__":
    if len(sys.argv) < 2:
        print("Usage: {} <filename>".format(sys.argv[0]), file=sys.stderr)
        sys.exit(1)

    try:
        with open(sys.argv[1], "rt") as f:
            mymap = make_map(f.read())
    except:
        print("Cannot open {}".format(sys.argv[1]), file=sys.stderr);
        sys.exit(1)

    print("Part1:", count_trees(mymap, (3, 1)))

    slopes = [(1,1), (3,1), (5,1), (7,1), (1,2)]
    print("Part2:", mul_slopes(mymap, slopes))
